stl/inc/flat_set

```diff
@@ -482,94 +481,102 @@ private:
+                // _Val <= *(_Where-1) ~ lower_bound is in [_Begin,_Where-1]
+                _Where = _STD lower_bound(_Begin, _Where - 1, _Val, _Pass_comp());
             }
         } else {
-            // look for the lower_bound for flat_set
-            if (_Where == _End || !_Compare(*_Where, _Val)) {
-                // _Val <= *_Where
-                if (_Where == _Begin || _Compare(*(_Where - 1), _Val)) {
-                    // _Val > *(_Where-1) ~ lower_bound is _Where
-                } else {
-                    // _Val <= *(_Where-1) ~ lower_bound is in [_Begin,_Where-1]
-                    _Where = _STD lower_bound(_Begin, _Where - 1, _Val, _Comp);
-                }
-            } else {
-                // _Val > *_Where ~ lower_bound is in [_Where+1,_End]
-                _Where = _STD lower_bound(_Where + 1, _End, _Val, _Comp);
-            }
+            // _Val > *_Where ~ lower_bound is in [_Where+1,_End]
+            _Where = _STD lower_bound(_Where + 1, _End, _Val, _Pass_comp());
         }
 
-        if constexpr (_Multi) {
-            _STL_INTERNAL_STATIC_ASSERT(is_same_v<remove_cvref_t<_Ty>, _Kty>);
-            _STL_INTERNAL_CHECK(_Check_where(_Where, _Val));
-            return _Cont.emplace(_Where, _STD forward<_Ty>(_Val));
-        } else {
-            if (_Where != _End && !_Compare(_Val, *_Where)) {
-                return _Cont.begin() + (_Where - _Begin);
-            }
+        if (_Where != _End && !_Compare(_Val, *_Where)) {
+            // convert const_iterator to iterator
+            return _Mycont.begin() + (_Where - _Begin);
+        }
 
-            if constexpr (is_same_v<remove_cvref_t<_Ty>, _Kty>) {
-                _STL_INTERNAL_CHECK(_Check_where(_Where, _Val));
-                return _Cont.emplace(_Where, _STD forward<_Ty>(_Val));
-            } else {
-                // for flat_set::insert(hint,auto&&)
-                _STL_INTERNAL_STATIC_ASSERT(_Keylt_transparent && is_constructible_v<_Kty, _Ty>);
-                _Kty _Keyval(_STD forward<_Ty>(_Val));
-                _STL_ASSERT(_Check_where(_Where, _Keyval), "The input type was not equivalent to key_type!");
-                return _Cont.emplace(_Where, _STD move(_Keyval));
-            }
+        if constexpr (is_same_v<remove_cvref_t<_Ty>, _Kty>) {
+            _STL_INTERNAL_CHECK(_Can_insert(_Where, _Val));
+            return _Mycont.emplace(_Where, _STD forward<_Ty>(_Val));
+        } else {
+            // heterogeneous insertion
+            // TRANSITION, the requirements may be stricter than needed by the standard.
+            _STL_INTERNAL_STATIC_ASSERT(_Keylt_transparent && is_constructible_v<_Kty, _Ty>);
+            _Kty _Keyval(_STD forward<_Ty>(_Val));
+            _STL_ASSERT(_Can_insert(_Where, _Keyval), "The conversion from the heterogeneous key to key_type should "
+                                                      "be consistent with the heterogeneous lookup!");

```

Comment:
>But we believe that the most common use-case for both heterogeneous lookup and insertion is the case when key_type and heterogeneous key represents the same thing like for std::string and std::string_view, which gives correct and unambiguous conversion from the heterogeneous key to key_type.

https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2021/p2363r2.html#dec_conv_comp_consistency

The current implementation works well when `key_type and heterogeneous key represents the same thing`, and the assertion implies this.

It's possible to support weirder (and much less useful) cases (and relax the assertion to only `!contains(_Keyval)`), without breaking the invariant at release mode, by forwarding the `_Keyval` to another `_Emplace_hint` function. This is, however, unfair for common cases, and will make the performance a lot worse. 

What does the standard say about `Conversion and comparison consistency`?